### ISSUE TITLE
Admin Calendar (based on BYU functionality)

### ISSUE THREAD

**Issue by jimhubert** [SFDO-Community/Summit-Events-App#261]
From: Matthew Gardiner
Sent: Friday, December 11, 2020 4:37 PM
To: Jim Hubert; Dahlberg, Thaddaeus B.
Subject: Re: [External] Summit Sprint
 
 would love to help get an admin calendar into Summit Events. It's pretty basic at this point. It simply allows you to see any Summit Event Instances, or instances that you own (when you click that button) via the owner custom property on a Summit Event Instance. There may be a better way to do that such as a one-to-many relationship instead of a one-to-one. You can drag and drop calendar events to change their time or length and that's pretty much it. I built a new Apex controller around the functionality but I suspect a lot of it is pretty similar to the Summit Events Feed class but I wanted to separate the controllers for public vs private access.

The real magic is that when you're doing AJAX to a Salesforce controller page, you can send an authentication header from the JavaScript including an authentication token that makes the Apex class treat the AJAX request as the currently logged in user. If you're using the browser native XMLHttpRequest that looks like this:

 xhttp = new XMLHttpRequest();

 xhttp.setRequestHeader('Authorization', 'Bearer {!$Api.Session_ID}');

I'll work on getting a github repository clone set up sometime next week. I'm not super familiar with github so I've been hesitant to do too much there. There are some byu specific paths that will have to be made generic I'm assuming by getting url paths in a more dynamic way.

Thanks,
Matt Gardiner

From: Jim Hubert <jim.hubert@fdnpartners.com>
Sent: Thursday, December 10, 2020 11:58 AM
To: Dahlberg, Thaddaeus B.; Matthew Gardiner
Subject: Re: [External] Summit Sprint
 
Matt and Thad,

Hope the two of you have had a chance to connect to discuss this, I agree that this calendar would be great to have in the Open Source version!

Jim Hubert  | Foundation Partners  |  763.232.1758  |  jim.hubert@fdnpartners.com 

From: Dahlberg, Thaddaeus B. <thad.dahlberg@stthomas.edu>
Sent: Thursday, December 3, 2020 2:14 PM
To: Matthew Gardiner <Matthew_Gardiner@byu.edu>; Jim Hubert <jim.hubert@fdnpartners.com>
Subject: RE: [External] Summit Sprint
 
Hi Matt,

That is a very cool calendar. We should get that in to the open source version!

There is currently no user relationship wired into the instances, but it is a good idea. Hanging off the instances there is a host object that is poorly implemented. We used it in the UST Events version to apply hosts to appointments. Like when we would have overnights the dorm host would be applied to the overnight appointment, or lunch would have the host of the person who you were going to eat lunch with at the event. It definitely needs to be architected better and perhaps we could add the event host into the mix. The trouble is where all these hosts live. Are they elevated to an equal level with the event objects so all objects can get at them. Perhaps I’m being confusing.

Let me know if you want to take some time to talk through some ideas.

Thanks

Thad

From: Matthew Gardiner <Matthew_Gardiner@byu.edu> 
Sent: Thursday, December 3, 2020 1:35 PM
To: Jim Hubert <jim.hubert@fdnpartners.com>; Dahlberg, Thaddaeus B. <thad.dahlberg@stthomas.edu>
Subject: [External] Summit Sprint

Hello,

I have a question that I'm hoping someone with a little more Salesforce experience can answer. I built a calendar that allows administrators to drag and drop event instances to reschedule them (or click them to go to the instance edit page). I've been asked to create a feature that takes the currently logged in user and only shows them events for which they are marked the host (the 'My Events' button in the screenshot below). The way that I've accomplished this is by having a lookup field on Instances to the User object called 'Owner' that a Salesforce user can set when creating/updating the instance. If there is a better way to do that without using a custom field I'd love to know. On initial inspection I didn't see any other relationship or object that served that purpose. In our case hosts for events will almost always be Admissions outreach folks and one of them spearheads the event even if other presenters are involved.

​Thanks,
Matt Gardiner
BYU Enrollment Services

**Comment by jimhubert** [comment]
[Screenshot of BYU Admin Calendar](https://drive.google.com/file/d/1bsDI8aSy3qofDXQp_A7K5GGV4YP2ag6h/view?usp=sharing)

**Comment by GardinerMatt** [comment]
I'm not sure if there is still a need for this since I've been out of the loop for so long but I'm back into doing Salesforce development full-time for the foreseeable future.  I think I'm going to convert it to a lightning component, if I can. I saw another issue that mentioned preferring lightning components to visualforce.

The way that I've implemented this is creating a custom object Summit_Event_Instance_Subscriber__c which has a master/detail relationship with Summit_Event_Instance__c and a lookup field for a user. Any user that is added as a subscriber can then see the event on their personal Calendar or all events if they click that tab. When a user creates a Summit Event Instance, they are automatically subscribed to it.

Let me know if anybody doesn't like how that's implemented. Happy to do it the way that works best for the project.

Anyway, unless I hear differently I'm going to go ahead with this approach and create a pull request when it's ready.
